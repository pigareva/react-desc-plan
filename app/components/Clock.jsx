import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Greeting from './Greeting';

export default class Clock extends Component {
  constructor(props) {
    super(props);
    this.state = { time: null };
  }

  componentDidMount() {
    this.startTime();
  }

  componentWillUnmount() {
    this.stopTime();
  }

  tick() {
    const currentTime = this.state.time || this.props.startTime;
    if (!this.props.isOff) {
      this.setState({
        time: currentTime >= this.props.endTime ? this.props.startTime : currentTime + 1,
      });
    }
  }

  stopTime() {
    clearInterval(this.timerID);
  }

  startTime() {
    this.timerID = setInterval(
      () => this.tick(),
      100,
    );
  }

  render() {
    return (
      <div>
        <div className="clock">{String(Math.floor(this.state.time / 60)).padStart(2, '0')}:{String(this.state.time % 60).padStart(2, '0')}</div>
        {this.props.isGreetingNeeded && <Greeting time={this.state.time} />}
      </div>
    );
  }
}

Clock.propTypes = {
  startTime: PropTypes.number,
  endTime: PropTypes.number,
  isOff: PropTypes.bool,
  isGreetingNeeded: PropTypes.bool,
};

Clock.defaultProps = {
  isGreetingNeeded: false,
  isOff: false,
  startTime: 0,
  endTime: 99999999999,
};
