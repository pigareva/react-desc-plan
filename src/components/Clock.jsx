import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class Clock extends Component {
  constructor(props) {
    super(props);
    this.state = { time: props.time, isOff: false };
    this.isGreetingNeeded = props.isGreetingNeeded || false;
  }

  componentDidMount() {
    this.startTime();
  }

  componentWillUnmount() {
    this.stopTime();
  }

  tick() {
    const currentTime = this.state.time;
    if (!this.state.isOff) {
      this.setState({ time: currentTime >= 1440 ? 540 : currentTime + 1 });
    }
  }

  stopTime() {
    clearInterval(this.timerID);
  }

  startTime() {
    this.timerID = setInterval(
      () => this.tick(),
      100,
    );
  }

  render() {
    return (
      <div className="clock">
        <p>{String(Math.floor(this.state.time / 60)).padStart(2, '0')}:{String(this.state.time % 60).padStart(2, '0')}</p>
      </div>
    );
  }
}

Clock.propTypes = {
  time: PropTypes.number.isRequired,
  isGreetingNeeded: PropTypes.bool,
};

Clock.defaultProps = {
  isGreetingNeeded: false,
};
