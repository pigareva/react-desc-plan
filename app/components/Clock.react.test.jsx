import React from 'react';
import renderer from 'react-test-renderer';
import Clock from './Clock';

jest.mock('./Greeting', () => 'Greeting');

test('Clock displays the time and the greeting', () => {
  const component = renderer.create(<Clock time={65} isGreetingNeeded />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

