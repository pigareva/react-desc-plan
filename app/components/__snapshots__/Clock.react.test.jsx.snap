// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clock displays the time and the greeting 1`] = `
<div>
  <div
    className="clock"
  >
    01
    :
    05
  </div>
  <Greeting
    time={65}
  />
</div>
`;
